scio-test/src/main/scala/com/spotify/scio/testing/CoderAssertions.scala

```diff
@@ -30,102 +32,185 @@ import scala.reflect.ClassTag
+  }
+
+  trait AssertionContextBase {
+    type ValType
+    val actualValue: Option[ValType]
+    val coder: Coder[ValType]
+    val lastAssertion: Option[Assertion]
+    val opts: PipelineOptions
+    lazy val beamCoder: BCoder[ValType] = CoderMaterializer.beamWithDefault(coder, opts)
+  }
+
+  def roundtrip[T: Equality](): CoderAssertionT[T] = ctx =>
+    checkRoundtripWithCoder[T](ctx.beamCoder, ctx.actualValue.get)
+
+  def roundtripToBytes[T: Equality](expectedBytes: Array[Byte]): CoderAssertionT[T] = ctx =>
+    checkRoundtripWithCoder[T](ctx.beamCoder, ctx.actualValue.get, expectedBytes)
+
+  def haveCoderInstance(expectedCoder: Coder[_]): CoderAssertion = ctx =>
+    ctx.coder should ===(expectedCoder)
+
+  def notFallback[T: ClassTag: Equality](): CoderAssertionT[T] = ctx => {
+    ctx.coder should !==(Coder.kryo[T])
+    checkRoundtripWithCoder(ctx.beamCoder, ctx.actualValue.get)
+  }
+
+  def fallback[T: ClassTag: Equality](): CoderAssertionT[T] = ctx => {
+    ctx.coder should ===(Coder.kryo[T])
+    checkRoundtripWithCoder(ctx.beamCoder, ctx.actualValue.get)
+  }
 
-  def notFallback[T: ClassTag](opts: PipelineOptions = DefaultPipelineOptions): CoderAssertion[T] =
-    new CoderAssertion[T] {
-      override def assert(value: T)(implicit c: Coder[T], eq: Equality[T]): Assertion = {
-        c should !==(Coder.kryo[T])
-        val beamCoder = CoderMaterializer.beamWithDefault(c, o = opts)
-        checkRoundtripWithCoder[T](beamCoder, value)
-      }
+  def beConsistentWithEquals(): CoderAssertion = ctx =>
+    ctx.beamCoder.consistentWithEquals() shouldBe true
+
+  def beNotConsistentWithEquals(): CoderAssertion = ctx =>
+    ctx.beamCoder.consistentWithEquals() shouldBe false
+
+  def beDeterministic(): CoderAssertion = ctx =>
+    noException should be thrownBy ctx.beamCoder.verifyDeterministic()
+
+  def beNonDeterministic(): CoderAssertion = ctx =>
+    a[NonDeterministicException] should be thrownBy ctx.beamCoder.verifyDeterministic()
+
+  def beSerializable(): CoderAssertion = ctx =>
+    noException should be thrownBy SerializableUtils.ensureSerializable(ctx.beamCoder)
+
+  def coderIsSerializable[A](implicit c: Coder[A]): Assertion =
+    c.coderShould(beSerializable()).lastAssertion.get
+
+  def beOfType[ExpectedCoder: ClassTag]: CoderAssertion = ctx => ctx.coder shouldBe a[ExpectedCoder]
+
+  def materializeTo[ExpectedBeamCoder: ClassTag]: CoderAssertion =
+    ctx => {
+      ctx.beamCoder shouldBe a[MaterializedCoder[_]]
+      ctx.beamCoder.asInstanceOf[MaterializedCoder[_]].bcoder shouldBe a[ExpectedBeamCoder]
     }
 
-  def fallback[T: ClassTag](opts: PipelineOptions = DefaultPipelineOptions): CoderAssertion[T] =
-    new CoderAssertion[T] {
-      override def assert(value: T)(implicit c: Coder[T], eq: Equality[T]): Assertion = {
-        c should ===(Coder.kryo[T])
-        roundtripKryo(opts).assert(value)
-      }
+  def materializeTo(className: String): CoderAssertion =
```

Comment:
Why is this needed ? is there some shortcoming with using the type parameter  `def materializeTo[ExpectedBeamCoder: ClassTag]` ?